### PR TITLE
[TEP-0135] Coschedule per (Isolated) PipelineRun e2e support

### PR DESCRIPTION
**Pull request by QuanZhang-William**
# Changes
Part of [#6740]. [TEP-0135][tep-0135] introduces a feature that allows a cluster operator to ensure that all of a PipelineRun's pods are scheduled to the same node.

This commit consumes the functions added in [#6819] and implements end to end support of `coschedule:pipelineruns` coscheduling mode, where all the `PipelineRun pods` are scheduled to the same node.

/kind feature

[#6819]: https://github.com/tektoncd/pipeline/pull/6819
[#6740]: https://github.com/tektoncd/pipeline/issues/6740
[tep-0135]: https://github.com/tektoncd/community/blob/main/teps/0135-coscheduling-pipelinerun-pods.md

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[TEP-0135]: Support `coschedule: pipelineruns` and `coschedule: isolate-pipelinerun` coschedule modes.
Users can now opt in this new feature to schedule all the pods in the same node and to optionally enforce one running pipelinerun in a node at the same time.
```

### CODE REVIEW

**Comment by QuanZhang-William** [comment]
It could make more sense to merge https://github.com/tektoncd/pipeline/pull/6929 first, so that this PR can have e2e supports for both `coschedule: pipelineruns` and `coschedule: isolate-pipelinerun` modes.

/cc @lbernick

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:91]: nit: you can use %s with `err.Error` rather than ignoring a linter

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:208]: I think we also need to delete the PVCs created by the statefulsets, and update integration tests to ensure those PVCs are actually deleted.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:227]: should this replace any of the functionality here? https://github.com/tektoncd/pipeline/blob/main/pkg/workspace/affinity_assistant_names.go

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:164]: nit: It doesn't really make sense to differentiate between nil and a pointer to an empty struct; I'd have tests treat them as equivalent rather than asserting a specific behavior.

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:370]: nit: could the switch statement be replaced with:
```
c.KubeClientSet.CoreV1().(*fake.FakeCoreV1).PrependReactor("create", tc.failureType,
				func(action testing2.Action) (handled bool, ret runtime.Object, err error) {
					return true, <something implementing runtime.Object>, errors.New("error creating %s", tc.failureType)
```

(I don't think the type of object returned in the nested func matters, but I could be wrong)

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:395]: nit: I think you can compare expectedErr and err directly

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:804]: Isn't this only true for affinity assistant per workspace?

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant_test.go:757]: why remove these tests?

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (lbernick) [inline pkg/reconciler/taskrun/taskrun_test.go:3352]: nit: can this be replaced with a more informative message? (or even just t.Fatal(err))

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (lbernick) [inline pkg/reconciler/taskrun/taskrun_test.go:3370]: could this be simplified to use existing helpers like t.IsFailure?

File: test/affinity_assistant_test.go
Comment (lbernick) [inline test/affinity_assistant_test.go:248]: This is interesting-- how do you prevent the integration test from interfering with the other integration tests? Do they just run sequentially?

File: test/affinity_assistant_test.go
Comment (lbernick) [inline test/affinity_assistant_test.go:252]: is there any way you could use the variables representing default values here rather than hard-coding them, so if we change the default values we don't also have to update this test?

**Comment by lbernick** [comment]
Thanks @QuanZhang-William! One more note; I think the existing release note might be hard for users to understand, and they might not know why it might be a good idea to enable this new feature. Can you put the release notes in terms of the functionality the user is interested in, i.e. scheduling pods to the right nodes?

Also, are all the docs for this feature up to date? I wouldn't want to have a release note saying the feature is ready but have our docs say it's not functional yet.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:208]: Thanks Lee! I have created a separate PR: https://github.com/tektoncd/pipeline/pull/6940 to deal with the cleanup logic (as well as the PVC deletion behaviors) as we discussed earlier.

I think it it makes more sense to merged https://github.com/tektoncd/pipeline/pull/6940 first, and I can add integration test in this PR

/hold

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:91]: Seems like change to %s still give me `errorlint` when running `golangci-lint`:

``` go			
return fmt.Errorf("%w: %s", ErrPvcCreationFailed, err.Error())
```

```
golangci-lint run --modules-download-mode=vendor --max-issues-per-linter=0 --max-same-issues=0 --deadline 5m

pkg/reconciler/pipelinerun/affinity_assistant.go:91:54: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
                        return fmt.Errorf("%w: %s", ErrPvcCreationFailed, err.Error())
```

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:237]: No I don't think so. This helper function is only used to calculate the value of `pipeline.tekton.dev/affinity-assistant` annotation

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:370]: hmmm, seems like the type need to match otherwise:

```
--- FAIL: TestCreateOrUpdateAffinityAssistantsAndPVCs_Failure (0.00s)
panic: interface conversion: runtime.Object is *v1.StatefulSet, not *v1.PersistentVolumeClaim [recovered]
	panic: interface conversion: runtime.Object is *v1.StatefulSet, not *v1.PersistentVolumeClaim
```

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:757]: function `isAffinityAssistantDisabled` is now no longer used and removed

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:164]: Prior to this commit, we don't create `AffinityAssistant` for `TaskRuns` without a `pvc` based workspace in `coschedule per pipelinerun` mode, the `statefulset` is NOT expected to be created (`nil`).

In this commit, we create `AffinityAssistant` for all `TaskRuns`, so we do expect an `AffinityAssitant` is created (but there is no `pvc` or `volumeclaimtemplate` in the `StatefulSetSpec`)

File: test/affinity_assistant_test.go
Comment (QuanZhang-William) [reply test/affinity_assistant_test.go:231]: Yeah, I'm following the discussion here: https://github.com/tektoncd/pipeline/issues/6079 to set/revert the feature flags and run tests sequentially.

**Comment by QuanZhang-William** [comment]
https://github.com/tektoncd/pipeline/pull/6940 and https://github.com/tektoncd/pipeline/pull/6929 are merged

/hold cancel

File: test/affinity_assistant_test.go
Comment (QuanZhang-William) [reply test/affinity_assistant_test.go]: Thanks! Changed to default value variables

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/taskrun/taskrun_test.go]: Changed to `t.IsFailure`

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/taskrun/taskrun_test.go]: error message updated

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go]: The cleanup logic test is updated in https://github.com/tektoncd/pipeline/pull/6940

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:396]: Here I'm validating the error messages so using `.Error()` (Error type is validated earlier).

It seems like I cannot compare error directly due to [`panic: cannot handle unexported field at {*fmt.wrapError}`](https://github.com/golang/go/blob/ffd9bd7e605cdd2eb66e38bad6e0d93b0d37963c/src/fmt/errors.go#L54C1-L57C2).

**Comment by QuanZhang-William** [comment]
>Thanks @QuanZhang-William! One more note; I think the existing release note might be hard for users to understand, and they might not know why it might be a good idea to enable this new feature. Can you put the release notes in terms of the functionality the user is interested in, i.e. scheduling pods to the right nodes?
> 

No problem, Lee. I have updated the release not

> Also, are all the docs for this feature up to date? I wouldn't want to have a release note saying the feature is ready but have our docs say it's not functional yet.

We have this open PR https://github.com/tektoncd/pipeline/pull/6892 for the documentation. I will remove all the WIP and Warnings in the documentation PR after merging this one.

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:396]: you can do this with cmpopts.EquateErrors I think? This will check that the error is the right type which is more important than validating the exact error message

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:812]: sorry, missed this!

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:163]: I'm a bit confused, why would we want to validate that a statefulset is created with an empty spec? how is that different from a nil pointer?

**Review by lbernick — APPROVED** [review]

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:396]: The type is validate here: https://github.com/tektoncd/pipeline/blob/e0d116852594ed0cf4106da4f2e40fbee2d47b26/pkg/reconciler/pipelinerun/affinity_assistant_test.go#L386C1-L395C4

To be considered equivalent in `cmpopts.EquateErrors()`, they must be the same error object so the validation will fail here.

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go]: Sorry for the confusion. The confusing point is that we had a `StatefulSetSpec` filter here: https://github.com/tektoncd/pipeline/blob/53e71bf8a014b6bb9174b8496bc3a829c700dd60/pkg/reconciler/pipelinerun/affinity_assistant_test.go#L1218, which ignores the `Replica` and `Selector` before.

In the test case, it is actually a spec with `Replica` and `Selector` but not an empty spec (we didn't have it before since they will not be validated anyways due to the filter). I think the best way address this confusion is to remote this filter and validate the selector and replicas (and we get more validated coverage 😄 )

File: pkg/reconciler/pipelinerun/affinity_assistant_test.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant_test.go:463]: I think cmpopts.EquateErrors uses errors.Is, which is what you're using in the check you linked? In that case you can remove the error string comparison, since I don't know that it adds much value compared to the error type and makes the tests more brittle. This is also fine to leave as is though.

**Comment by QuanZhang-William** [comment]
/hold cancel
v0.50 LTS is released

**Pull request merged**
